### PR TITLE
Fix: Web Socket Invocation Error with Tracing Enabled

### PR DESCRIPTION
**Pull request by BLasan**
### Purpose
$subject

### Issue
- https://github.com/wso2/api-manager/issues/1550

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/common/APIMgtLatencySynapseHandler.java [hunk 1]
```diff
@@ -129,14 +136,16 @@ public boolean handleResponseOutFlow(MessageContext messageContext) {
...
-                TelemetryUtil.updateOperation(responseLatencySpan,
-                        api.getApiName().concat("--").concat(api.getApiVersion()).concat("--")
-                                .concat(GatewayUtils.getTenantDomain()));
+            if (responseLatencySpan != null) {
+                GatewayUtils.setAPIRelatedTags(responseLatencySpan, messageContext);
+                API api = GatewayUtils.getAPI(messageContext);
+                if (api != null) {
+                    TelemetryUtil.updateOperation(responseLatencySpan,
+                            api.getApiName().concat("--").concat(api.getApiVersion()).concat("--")
+                                    .concat(GatewayUtils.getTenantDomain()));
```
Comment (senthuran16) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/common/APIMgtLatencySynapseHandler.java:145]: Let's extract this into a separate method. There is more than one usage.

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/SubscriptionDataHolder.java [hunk 2]
```diff
@@ -65,16 +65,19 @@ public void unregisterTenantSubscriptionStore(String tenantDomain) {
...
-        SubscriptionDataStore subscriptionDataStore = subscriptionStore.get(tenantDomain);
-        if (subscriptionDataStore == null) {
-            synchronized (tenantDomain.concat("getTenantSubscriptionStore").intern()) {
-                subscriptionDataStore = subscriptionStore.get(tenantDomain);
-                if (subscriptionDataStore == null) {
-                    subscriptionDataStore = registerTenantSubscriptionStore(tenantDomain);
+        if (subscriptionStore != null && tenantDomain != null) {
+            SubscriptionDataStore subscriptionDataStore = subscriptionStore.get(tenantDomain);
+            if (subscriptionDataStore == null) {
+                synchronized (tenantDomain.concat("getTenantSubscriptionStore").intern()) {
```
Comment (senthuran16) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/SubscriptionDataHolder.java:71]: Let's define a constant for `getTenantSubscriptionStore` and use here.

**Review by senthuran16 — CHANGES_REQUESTED** [review]

**Review by senthuran16 — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/SubscriptionDataHolder.java [hunk 3]
```diff
@@ -65,16 +65,19 @@ public void unregisterTenantSubscriptionStore(String tenantDomain) {
...
+                    subscriptionDataStore = subscriptionStore.get(tenantDomain);
+                    if (subscriptionDataStore == null) {
+                        subscriptionDataStore = registerTenantSubscriptionStore(tenantDomain);
+                    }
                 }
             }
+            return subscriptionDataStore;
         }
-        return subscriptionDataStore;
+        return null;
```
Comment (Arshardh) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/SubscriptionDataHolder.java:80]: Did we check if there are any possible null pointers that can occur as a result of sending null.

Comment (BLasan) [reply components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/SubscriptionDataHolder.java:80]: When analyzing the master branch, line 80 shouldn't get hit. This `null` check is added in https://github.com/wso2/carbon-apimgt/blob/master/components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/AbstractKeyValidationHandler.java#L219

**Pull request merged**